Context support for GUI-driven clients and relays

abakum opened the pull request:
>https://github.com/schollz/croc/issues/943
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
...
c, err:=NewCtx(ctx, Options{...})
...
go func() {
 c.Send(...)
 donechan <- true
}()
go func() {
 <-cancelchan
 cancel()
}()
...
go func() {
 c.Receive()
 donechan <- true
}()
go func() {
 <-cancelchan
 cancel()
}()
...

abakum committed:
Context support for GUI-driven clients and relays

abakum committed:
Context cancellation for hashing.

abakum committed:
Use LocalIP() on error in GetLocalIPs() helps for hosts with android

abakum committed:
updateIfSenderChannelSecured

abakum commented:
This code works. Zack, please try it out to believe it. https://github.com/abakum/crocgui

abakum committed:
feat(tcp): allow default password clients to join existing rooms but not create new ones

Move password validation from early in clientCommunication to only when creating a new room. This enables clients with the default password to join existing rooms (e.g., as receivers) while still preventing them from creating rooms, thus maintaining security for room creation. Add tests to verify:
- Default password can join existing room.
- Default password cannot create new room.
- Correct password retains full access.

schollz commented:
@abakum if you want to PR https://github.com/schollz/croc/pull/1038/commits/cf05c8285fd9f537e96085171e07e4e588437ba2, please make it in a clean PR

abakum commented:
Ok.

abakum committed:
fix: reject arbitrary wrong passwords immediately

Clients with passwords that are neither correct nor the default passphrase are now rejected right away, instead of being allowed to attempt room creation or joining. This prevents unauthorized clients from enumerating rooms or causing unnecessary load.

Test added to verify early rejection of arbitrary wrong passwords.

abakum committed:
fix: handle connection failures gracefully and protect against brute-force password attempts

- Replace panic with error logging and clean return on connection failure in croc/croc.go
- Add 1-second delay before rejecting incorrect passwords in tcp.go to rate-limit brute-force attacks

abakum commented:
Wow! https://github.com/schollz/croc/pull/1060 Continue with https://github.com/schollz/croc/pull/1106
